app/parse.py

```diff
@@ -15,9 +29,104 @@ class Product:
     num_of_reviews: int
 
 
+def scrape_single_product(product: WebElement) -> Product:
+    title = product.find_element(By.CLASS_NAME, "title").get_attribute("title")
+    price, _, description, num_of_reviews = product.text.split("\n")
+    price = float(price.replace("$", ""))
+    num_of_reviews = int(num_of_reviews.split()[0])
+
+    rating = (
+        product.find_element(By.CLASS_NAME, "ratings")
+        .find_elements(By.TAG_NAME, "p")[-1]
+        .get_attribute("data-rating")
+    )
+
+    rating = int(rating) if rating else 5
+
+    return Product(
+        title=title,
+        price=price,
+        description=description,
+        num_of_reviews=num_of_reviews,
+        rating=rating,
+    )
+
+
+def scrape_page(url: str, driver: WebDriver) -> list[Product]:
+    driver.get(url)
+
+    try:
+        if accept := driver.find_element(By.CLASS_NAME, "acceptCookies"):
+            time.sleep(0.5)
+            accept.click()
+
+    except NoSuchElementException:
+        pass
+
+    try:
+        while more := driver.find_element(
+                By.CLASS_NAME, "ecomerce-items-scroll-more"
+        ):
+            if not more.is_displayed():
+                break
+            more.click()
+
+    except (
+        NoSuchElementException,
+        ElementNotInteractableException,
+        ElementClickInterceptedException,
+    ):
+        pass
+
+    products = driver.find_elements(By.CLASS_NAME, "card-body")
+
+    return [scrape_single_product(product) for product in products]
+
+
+def get_urls(class_name: str, driver: WebDriver) -> list[tuple]:
+    return [
+        (elem.text.lower(), elem.get_attribute("href"))
+        for elem in driver.find_elements(By.CLASS_NAME, class_name)
+    ]
+
+
+def get_sidebar_urls(home_url: str, driver: WebDriver) -> list[tuple | tuple[str, str]]:
```

Comment:
The return type hint `list[tuple | tuple[str, str]]` is inconsistent. It suggests that the list can contain tuples of different structures, which might be confusing. Consider using a consistent tuple structure or clarifying the type hint.